app/controlplane/pkg/data/attestationstate.go

```diff
@@ -48,26 +51,69 @@ func (r *AttestationStateRepo) Initialized(ctx context.Context, runID uuid.UUID)
 	return exists, nil
 }
 
-func (r *AttestationStateRepo) Save(ctx context.Context, runID uuid.UUID, state []byte) error {
-	err := r.data.DB.WorkflowRun.UpdateOneID(runID).SetAttestationState(state).Exec(ctx)
+// baseDigest, when provided will be used to check that it matches the digest of the state currently in the DB
+// if the digests do not match, the state has been modified and the caller should retry
+func (r *AttestationStateRepo) Save(ctx context.Context, runID uuid.UUID, state []byte, baseDigest string) (err error) {
```

Comment:
this method has the meat of the implementation, it makes sure that 

in a transaction

- Reads the state stored in DB and blocks the row for update
- Calculates the sha of its content and compares it with the baseDigest optionally provided
- if it doesn't match, it returns a ConflictError and cancels the transaction, unblocking the row.
- If it does, we update the column and unblock the row